### PR TITLE
Introduce NSX-T support for External Network

### PR DESCRIPTION
**Pull request by Didainius**
As work continues for NSX-T support this PR adds a new type `ExternalNetworkV2` for managing external networks using OpenAPI endpoint. This is required to handle NSX-T external networks.

In addition to this this PR adds query functions for NSX-T related resources - NSX-T managers and Tier-0 routers (`QueryNsxtManagerByName` and `GetImportableNsxtTier0RouterByName` respectively)

**Note.** Tests require additional configuration variables as per `govcd/sample_govcd_test_config.yaml `

### CODE REVIEW

File: govcd/api.go [hunk 1]
```diff
@@ -642,3 +643,46 @@ func takeIntAddress(x int) *int {
...
+// IsUuid returns true if the identifier is a bare UUID
+func IsUuid(identifier string) bool {
+	reUuid := regexp.MustCompile(`^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$`)
+	return reUuid.MatchString(identifier)
+}
+
+// isUrn validates if supplied identifier is of URN format (e.g. urn:vcloud:nsxtmanager:09722307-aee0-4623-af95-7f8e577c9ebc)
+// it checks for the following criteria:
+// 1. idenfifier is not empty
+// 2. identifier has 4 elements separated by ';'
```
Comment (vbauzysvmware) [inline govcd/api.go:656]:
```suggestion
// 2. identifier has 4 elements separated by ':'
```

File: govcd/external_network_v2.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+)
+
+// ExternalNetworkV2 is a type for version 2 of external network which uses OpenAPI endpoint to manage external networks
+type ExternalNetworkV2 struct {
+	ExternalNetwork *types.ExternalNetworkV2
+	client          *Client
+}
+
+// CreateExternalNetwork creates a new external network using OpenAPI endpoint
+func CreateExternalNetworkV2(vcdClient *VCDClient, newExtNet *types.ExternalNetworkV2) (*ExternalNetworkV2, error) {
```
Comment (vbauzysvmware) [inline govcd/external_network_v2.go:21]: why we don't make vcdClient the `recipient` ?

File: govcd/external_network_v2.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+)
+
+// ExternalNetworkV2 is a type for version 2 of external network which uses OpenAPI endpoint to manage external networks
+type ExternalNetworkV2 struct {
+	ExternalNetwork *types.ExternalNetworkV2
+	client          *Client
+}
+
+// CreateExternalNetwork creates a new external network using OpenAPI endpoint
```
Comment (vbauzysvmware) [inline govcd/external_network_v2.go:20]: I think it's worth to mentioned similar things like it is in terraform. Like it create nsx-v, nsx-t and which fields supported by which type.

File: govcd/external_network_v2_test.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	check.Assert(err, IsNil)
+
+	// Create network and test CRUD capabilities
+	netNsxt := testExternalNetworkV2(types.ExternalNetworkBackingTypeNsxtTier0Router, tier0Router.NsxtTier0Router.ID, nsxtManagerId)
+	createdNet, err := CreateExternalNetworkV2(vcd.client, netNsxt)
+	check.Assert(err, IsNil)
+
+	createdNet.ExternalNetwork.Name = "changed_name"
+	_, err = createdNet.Update()
+	check.Assert(err, IsNil)
```
Comment (vbauzysvmware) [inline govcd/external_network_v2_test.go:39]: updated named isn't checked

File: govcd/external_network_v2_test.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+package govcd
+
+import (
+	"fmt"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	. "gopkg.in/check.v1"
+)
+
+func (vcd *TestVCD) Test_CreateExternalNetworkV2NsxT(check *C) {
```
Comment (vbauzysvmware) [inline govcd/external_network_v2_test.go:16]: missing test Test_CreateExternalNetworkV2NsxV or it's name is `Test_CreateExternalNetworkV2PortGroup` ?

File: govcd/external_network_v2_test.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	check.Assert(err, IsNil)
+
+	neT := testExternalNetworkV2(vcd.config.VCD.ExternalNetworkPortGroupType, pgs[0].MoRef, vcUrn)
+
+	r, err := CreateExternalNetworkV2(vcd.client, neT)
+	check.Assert(err, IsNil)
+
+	r.ExternalNetwork.Name = "changed_name"
+	_, err = r.Update()
+	check.Assert(err, IsNil)
```
Comment (vbauzysvmware) [inline govcd/external_network_v2_test.go:102]: updated named isn't checked

File: govcd/nsxt_tier0_router.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+		return nil, fmt.Errorf("NSX-T manager ID is not URN (e.g. 'urn:vcloud:nsxtmanager:09722307-aee0-4623-af95-7f8e577c9ebc)', got: %s", nsxtManagerId)
+	}
+
+	if name == "" {
+		return nil, fmt.Errorf("empty Tier 0 router name specified")
+	}
+
+	// Ideally FIQL filter could be used to filter on server side and get only desired result, but filtering on
+	// 'displayName' is not yet supported. The only supported field for filtering is
+	// _context==urn:vcloud:nsxtmanager:09722307-aee0-4623-af95-7f8e577c9ebc to specify parent NSX-T manager (This
```
Comment (vbauzysvmware) [inline govcd/nsxt_tier0_router.go:43]: `_context` or `context` ?

File: govcd/nsxt_tier0_router.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+		queryParameters.Add("filter", "displayName=="+name)
+	*/
+
+	nsxtTier0Routers, err := vcdCli.GetAllImportableNsxtTier0Routers(nsxtManagerId, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not find NSX-T Tier-0 router with name '%s' for NSX-T manager with id '%s': %s",
+			name, nsxtManagerId, err)
+	}
+
+	// TODO remove this when FIQL supports filtering on 'displayName'
```
Comment (vbauzysvmware) [inline govcd/nsxt_tier0_router.go:56]: maybe better is move this part to private func and deprecate it :)

File: types/v56/openapi.go [hunk 9]
```diff
@@ -50,3 +50,69 @@ type Role struct {
...
+	DisplayName string `json:"displayName"`
+}
+
+// ExternalNetworkV2 defines a struct for OpenAPI endpoint
+type ExternalNetworkV2 struct {
+	ID              string                    `json:"id,omitempty"`
+	Name            string                    `json:"name"`
+	Description     string                    `json:"description"`
+	Subnets         ExternalNetworkV2Subnets  `json:"subnets"`
+	NetworkBackings ExternalNetworkV2Backings `json:"networkBackings"`
```
Comment (vbauzysvmware) [inline types/v56/openapi.go:67]: I think would be good to have comment with possible values

**Review by vbauzysvmware — CHANGES_REQUESTED** [review]
first look

File: govcd/external_network_v2.go (hunk 2 above)
Comment (Didainius) [reply govcd/external_network_v2.go:21]: Can do - the only reason is I tried to follow consistency. We already have `func CreateExternalNetwork(vcdClient *VCDClient, externalNetworkData *types.ExternalNetwork) (Task, error) {`

File: govcd/external_network_v2_test.go (hunk 5 above)
Comment (Didainius) [reply govcd/external_network_v2_test.go:16]: Yes. `Test_CreateExternalNetworkV2PortGroup`

File: govcd/nsxt_tier0_router.go (hunk 7 above)
Comment (Didainius) [reply govcd/nsxt_tier0_router.go:43]: Yes `_context`. Like a hidden var or whatever :)

Comment (vbauzysvmware) [reply govcd/nsxt_tier0_router.go:43]: I like it, one more nasty thing to bucket :+1:

File: govcd/external_network_v2.go (hunk 2 above)
Comment (vbauzysvmware) [reply govcd/external_network_v2.go:21]: @lvirbalas @dataclouder

File: govcd/external_network_v2_test.go (hunk 5 above)
Comment (vbauzysvmware) [reply govcd/external_network_v2_test.go:16]: so maybe better name it Test_CreateExternalNetworkV2NsxV ?:)

File: CHANGELOG.md [hunk 10]
```diff
@@ -22,6 +22,10 @@
 * Dropped support for VMware Cloud Director 9.5 [#330](https://github.com/vmware/go-vcloud-director/pull/330)
 * Deprecated Vdc.UploadMediaImage because it no longer works with API V32.0+ [#330](https://github.com/vmware/go-vcloud-director/pull/330)
 * Introduce NSX-T support for adminOrg.CreateOrgVdc() [#332](https://github.com/vmware/go-vcloud-director/pull/332)
+* Introduce NSX-T support for external network using OpenAPI endpoint and `ExternalNetworkV2` type methods including `CreateExternalNetworkV2`, 
+`GetExternalNetworkById`, `GetAllExternalNetworks`, `Update`, and `DELETE` [#335](https://github.com/vmware/go-vcloud-director/pull/335)
```
[the remark this answers is not in the document]
Comment (Didainius) [reply CHANGELOG.md]: Changed.

File: CHANGELOG.md [hunk 11]
```diff
@@ -22,6 +22,10 @@
...
+* Introduce NSX-T Query functions `QueryNsxtManagerByName` and `GetImportableNsxtTier0RouterByName` [#335](https://github.com/vmware/go-vcloud-director/pull/335)
```
[the remark this answers is not in the document]
Comment (Didainius) [reply CHANGELOG.md]: Changed.

File: govcd/api.go (hunk 1 above)
Comment (Didainius) [reply govcd/api.go]: Fixed.

File: govcd/external_network_v2.go (hunk 3 above)
Comment (Didainius) [reply govcd/external_network_v2.go]: Added more details

File: govcd/external_network_v2_test.go (hunk 4 above)
Comment (Didainius) [reply govcd/external_network_v2_test.go:51]: Fixed.

File: govcd/external_network_v2_test.go (hunk 5 above)
Comment (Didainius) [reply govcd/external_network_v2_test.go]: Can be. Done.

File: govcd/external_network_v2_test.go (hunk 6 above)
Comment (Didainius) [reply govcd/external_network_v2_test.go:115]: Fixed.

File: govcd/nsxt_tier0_router.go (hunk 8 above)
Comment (Didainius) [reply govcd/nsxt_tier0_router.go:56]: Moved to a function.

File: types/v56/openapi.go (hunk 9 above)
Comment (Didainius) [reply types/v56/openapi.go:74]: Added more docs whatever I could find

File: govcd/sample_govcd_test_config.yaml [hunk 12]
```diff
@@ -127,6 +127,13 @@ vcd:
       # The minimum size is 1048576 (1 MB). While theoretically smaller amounts are allowed,
       # there is an issue when using < 1MB during updates
       sizeForUpdate: 1048576
+nsxt:
+    # NSX-T manager to be used
+    manager: nsxManager1
+    # NSX-T tier-0 router used for external network
+    tier0router: tier-0-router
+    # NSX-T tier-0 VRF router used for external network
```
Comment (lvirbalas) [inline govcd/sample_govcd_test_config.yaml:135]: For those new to NSX-T would be nice to have a comment explaining how VRF router differs from the normal one.

**Review by lvirbalas — CHANGES_REQUESTED** [review]

File: govcd/sample_govcd_test_config.yaml (hunk 12 above)
Comment (Didainius) [reply govcd/sample_govcd_test_config.yaml:135]: It ties to down to VRF ([virtual routing and forwarding](https://en.wikipedia.org/wiki/Virtual_routing_and_forwarding)). I can't say yet I am an expert myself (at least yet) :) As I understand so far it allows to build virtual routing tables on the same Tier0 router, but I am yet to experience it better while building NSX-T from scratch.

I do know that from VCD standpoint these are differentiated and API requires different values to be passed so it makes sense to have explicit test for VRF Tier0 router.

File: govcd/external_network_v2.go (hunk 2 above)
Comment (lvirbalas) [reply govcd/external_network_v2.go:24]: No strong preference here, but like consistency :)

**Review by lvirbalas — APPROVED** [review]

File: govcd/external_network_v2_test.go [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	. "gopkg.in/check.v1"
+)
+
+func (vcd *TestVCD) Test_CreateExternalNetworkV2Nsxt(check *C) {
+	vcd.testCreateExternalNetworkV2Nsxt(check, vcd.config.VCD.Nsxt.Tier0router, types.ExternalNetworkBackingTypeNsxtTier0Router)
+}
+
+func (vcd *TestVCD) Test_CreateExternalNetworkV2NsxtVrf(check *C) {
+	// Checking NSX-T VRF router would be expected to require
```
Comment (vbauzysvmware) [inline govcd/external_network_v2_test.go:21]: Could you rephrase? - I didn't grasp idea of this message

File: govcd/api_vcd_test.go [hunk 14]
```diff
@@ -166,6 +166,11 @@ type TestConfig struct {
 			Size          int64 `yaml:"size,omitempty"`
 			SizeForUpdate int64 `yaml:"sizeForUpdate,omitempty"`
 		}
+		Nsxt struct {
+			Manager        string `yaml:"manager"`
+			Tier0router    string `yaml:"tier0router"`
+			Tier0routerVrf string `yaml:"tier0routerVrf"`
```
Comment (vbauzysvmware) [inline govcd/api_vcd_test.go:172]: in every config example `tier0routervrf` is used, but by code expected `tier0routerVrf`

**Review by vbauzysvmware — APPROVED** [review]
Just a few fix and should be ok. Also I am not fond of naming files with _V2.go - we should seek better naming. Maybe even _nsxt.go is ok enough. We can change that later I think we better thought on that.

File: govcd/api_vcd_test.go (hunk 14 above)
Comment (Didainius) [reply govcd/api_vcd_test.go:172]: Fixed.

File: govcd/external_network_v2_test.go (hunk 13 above)
Comment (Didainius) [reply govcd/external_network_v2_test.go]: Changed.

**Comment by Didainius** [comment]
> Just a few fix and should be ok. Also I am not fond of naming files with _V2.go - we should seek better naming. Maybe even _nsxt.go is ok enough. We can change that later I think we better thought on that.

Open to proposals and renaming. It can't be `_nsxt` though as this network supports NSX-V as well using OpenAPI.

**Comment by dataclouder** [comment]
> > Just a few fix and should be ok. Also I am not fond of naming files with _V2.go - we should seek better naming. Maybe even _nsxt.go is ok enough. We can change that later I think we better thought on that.
> 
> Open to proposals and renaming. It can't be `_nsxt` though as this network supports NSX-V as well using OpenAPI.

We had a discussion about it, and `_v2` was the winner

**Review by dataclouder — APPROVED** [review]
